File: secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import           Data.Bits (testBit)


data Event = Wink
           | DoubleBlink
           | CloseYourEyes
           | Jump
           | Reverse
           deriving (Enum, Bounded)


instance Show Event where
  show Wink          = "wink"
  show DoubleBlink   = "double blink"
  show CloseYourEyes = "close your eyes"
  show Jump          = "jump"
  show Reverse       = "reverse"


fromInt :: Int -> [Event]
fromInt n = [ event | event <- Reverse : [Wink .. Jump]
                    , n `testBit` fromEnum event ]


handshake :: Int -> [String]
handshake = foldr go [] . fromInt
  where
    go :: Event -> [String] -> [String]
    go Reverse = reverse
    go event   = (show event :)

File: erlang/secret-handshake/src/secret_handshake.erl
-module(secret_handshake).

-export([commands/1]).

-spec commands(non_neg_integer()) -> [string()].
commands(Number) when Number band 16 =:= 16 ->
    lists:reverse(commands(Number - 16));
commands(Number) when Number band 1 =:= 1 ->
    ["wink" | commands(Number - 1)];
commands(Number) when Number band 2 =:= 2 ->
    ["double blink" | commands(Number - 2)];
commands(Number) when Number band 4 =:= 4 ->
    ["close your eyes" | commands(Number - 4)];
commands(Number) when Number band 8 =:= 8 ->
    ["jump" | commands(Number - 8)];
commands(_) ->
    [].
